W 12364191 whats new dw v25 duke (#222)
mulesoft/docs-dataweave: modules/ROOT/pages/whats-new-in-dw.adoc

= What's New in DataWeave

DataWeave 2.5 is bundled with Mule 4.5. The 2.5 version of DataWeave introduces some new features.

== Versioning Behavior in DataWeave

> DataWeave now supports backward compatibility with previous 2.x versions of DataWeave through compatibility flags and syntax-level support:

Author: ```suggestion
DataWeave now enables backward compatibility with previous 2.x versions of DataWeave through compatibility flags and syntax-level support:
```

* Syntax-level support is available at the script level through the `%dw` directive (see xref:dataweave-language-introduction.adoc#dw_header[DataWeave Scripts]). Setting the directive to an earlier version of DataWeave avoids any syntax-breaking changes when the DataWeave runtime engine runs the script. This script-level setting enables you to retain earlier behavior in some scripts while using the latest behavior in others.
// 2.5 DOC MERGED: https://github.com/mulesoft/docs-dataweave/pull/214/files

For detail, see xref:dataweave-versioning-behavior.adoc[].

== Supported Formats

DataWeave now supports the ProtoBuf format and extends its XML support to DTD.

* Protocol Buffers (Protobuf): This format is typically used with gRPC, the Remote Procedure Call (RPC) framework. See xref:dataweave-formats-protobuf.adoc[].

* DTD for XML: DataWeave can read and create cXML. However, the DTDs remain disabled for security reasons. See xref:dw-xml.adoc[].
//TODO: Does this related to the parseDtd flag?
//TODO: How does this relate to compatibility flags?
//From Lische: https://salesforce-internal.slack.com/archives/C011SNL2469/p1671213207700189

> 

Author: TODO: To read a doctype declarations, set the xref:dataweave-system-properties.adoc[system property] `com.mulesoft.dw.xml_reader.parseDtd`.

Author: Also, `DocType` is the DW type for the doctype declaration, and `docType` is the variable containing the value of the doctype declaration in a `DocType` value.

Author: new module is Dtd (dw::xml::Dtd)

== Specifying Type Parameters

See "Specify Type Parameters" in xref:dataweave-functions-lambdas.adoc[].

* Typing enhancements that allow for reuse of type definitions from other sources:
** Type parameters in a function call.  //TODO: ADD anchored link to section.
See "Type Parameters Definition" in xref:dataweave-functions.adoc[].
//TODO: ADD anchored link to section.
//From ANA: Type parameters in function calls (@Andrés Radunsky): This is a very
//          advanced scenario where typed parameters, or generics, will now be able
//          to be declared at the function call level.  So one can say
//          myFunc<String>()  to signal that the generic in myFunc<T>() should
//          be a String. In the past, generics were always inferred which led
//          to some type checking errors. Because this required a syntax change,
//          it will only be available to scripts declaring %dw 2.5  or higher
//          (in the future) as their desired version.
//github MERGE (type parameter application):
//             https://github.com/mulesoft/docs-dataweave/pull/215/files

> 

Author: TODO: NEW functions, formats, types, modules

	dataweave-formats-protobuf.adoc
	dw-binaries-functions-concatwith.adoc
	dw-dtd-functions-doctypeasstring.adoc
	dw-dtd-types.adoc
	dw-dtd.adoc
	dw-runtime-functions-version.adoc

Author: modified:

	modified:   dw-binaries-functions-frombase64.adoc
	modified:   dw-binaries-functions-fromhex.adoc
	modified:   dw-binaries-functions-tobase64.adoc
	modified:   dw-binaries.adoc
	modified:   dw-coercions-functions-toarray.adoc
	modified:   dw-coercions-functions-tobinary.adoc
	modified:   dw-coercions-functions-todate.adoc
	modified:   dw-coercions-functions-todatetime.adoc
	modified:   dw-coercions-functions-tolocaldatetime.adoc
	modified:   dw-coercions-functions-tolocaltime.adoc
	modified:   dw-coercions-functions-tonumber.adoc
	modified:   dw-coercions-functions-tostring.adoc
	modified:   dw-coercions-functions-totime.adoc
	modified:   dw-core-annotations.adoc
	modified:   dw-core-functions-distinctby.adoc
	modified:   dw-core-functions-filter.adoc
	modified:   dw-core-functions-filterobject.adoc
	modified:   dw-core-functions-groupby.adoc
	modified:   dw-core-functions-minusminus.adoc
	modified:   dw-core-functions-typeof.adoc
	modified:   dw-core.adoc
	modified:   dw-crypto-functions-hmacwith.adoc
	modified:   dw-dataformat-types.adoc
	modified:   dw-dates-functions-atbeginningofday.adoc
	modified:   dw-dates-functions-atbeginningofhour.adoc
	modified:   dw-dates-functions-atbeginningofmonth.adoc
	modified:   dw-dates-functions-atbeginningofweek.adoc
	modified:   dw-dates-functions-atbeginningofyear.adoc
	modified:   dw-dates-functions-date.adoc
	modified:   dw-dates-functions-datetime.adoc
	modified:   dw-dates-functions-localdatetime.adoc
	modified:   dw-dates-functions-localtime.adoc
	modified:   dw-dates-functions-time.adoc
	modified:   dw-dates-functions-today.adoc
	modified:   dw-dates.adoc
	modified:   dw-math-functions-acos.adoc
	modified:   dw-math-functions-asin.adoc
	modified:   dw-math-functions-atan.adoc
	modified:   dw-math-functions-cos.adoc
	modified:   dw-math-functions-log10.adoc
	modified:   dw-math-functions-logn.adoc
	modified:   dw-math-functions-sin.adoc
	modified:   dw-math-functions-tan.adoc
	modified:   dw-math-functions-todegrees.adoc
	modified:   dw-math-functions-toradians.adoc
	modified:   dw-math-variables.adoc
	modified:   dw-math.adoc
	modified:   dw-multipart-functions-field.adoc
	modified:   dw-multipart-functions-file.adoc
	modified:   dw-multipart-functions-form.adoc
	modified:   dw-multipart.adoc
	modified:   dw-periods-functions-days.adoc
	modified:   dw-periods-functions-duration.adoc
	modified:   dw-periods-functions-hours.adoc
	modified:   dw-periods-functions-minutes.adoc
	modified:   dw-periods-functions-months.adoc
	modified:   dw-periods-functions-period.adoc
	modified:   dw-periods-functions-seconds.adoc
	modified:   dw-periods-functions-years.adoc
	modified:   dw-runtime-functions-eval.adoc
	modified:   dw-runtime-types.adoc
	modified:   dw-runtime.adoc
	modified:   dw-strings-functions-countcharactersby.adoc
	modified:   dw-strings-functions-countmatches.adoc
	modified:   dw-strings-functions-everycharacter.adoc
	modified:   dw-strings-functions-first.adoc
	modified:   dw-strings-functions-last.adoc
	modified:   dw-strings-functions-somecharacter.adoc
	modified:   dw-strings-functions-substringby.adoc
	modified:   dw-strings-functions-withmaxsize.adoc

Author: added

= What's New in DataWeave

DataWeave 2.5 is bundled with Mule 4.5. The 2.5 version of DataWeave introduces some new features.

== Versioning Behavior in DataWeave

> DataWeave now enables backward compatibility with previous 2.x versions of DataWeave at the Mule application and DataWeave script levels:

Author: ```suggestion
DataWeave enables backward compatibility with previous 2.x versions of DataWeave at the Mule application and DataWeave script levels:
```

* Compatibility flags are a set of DataWeave xref:dataweave-system-properties.adoc[DataWeave system properties] that enable you to retain specific DataWeave behavior at the Mule application level. The flags are valid for specified DataWeave versions and expire in a future version of DataWeave. Flags change behavior based on the value of the Mule `minMuleVersion` system property, which corresponds to a DataWeave version. See xref:feature-flagging.adoc[] for details.
//2.5 DOC MERGED: https://github.com/mulesoft/docs-dataweave/pull/214/files

* Syntax-level support is available at the script level through the `%dw` directive (see xref:dataweave-language-introduction.adoc#dw_header[DataWeave Scripts]). Setting the directive to an earlier version of DataWeave avoids any syntax-breaking changes when the DataWeave runtime engine runs the script. This script-level setting enables you to retain earlier behavior in some scripts while using the latest behavior in others.
// 2.5 DOC MERGED: https://github.com/mulesoft/docs-dataweave/pull/214/files

For detail, see xref:dataweave-versioning-behavior.adoc[].

== Supported Formats

> DataWeave supports the ProtoBuf format and extends its XML support to DTD.

Author: ```suggestion
DataWeave supports the ProtoBuf format and extends its XML support to reading and creating doctype directives (DTDs).
```

> DataWeave now supports the ProtoBuf format and extends its XML support to DTD.

Author: ```suggestion
DataWeave supports the ProtoBuf format and extends its XML support to DTD.
```

//          myFunc<String>()  to signal that the generic in myFunc<T>() should
//          be a String. In the past, generics were always inferred which led
//          to some type checking errors. Because this required a syntax change,
//          it will only be available to scripts declaring %dw 2.5  or higher
//          (in the future) as their desired version.
//github MERGE (type parameter application):
//             https://github.com/mulesoft/docs-dataweave/pull/215/files

** Creating new DataWeave types from existing types. The dot selector (`.`) enables you to navigate over DataWeave xref:dataweave-type-system.adoc#composite-types[type] definitions and define new types from existing ones. Combined with the existing xref:dataweave-selectors-reusing-types.adoc[loader] for DataWeave modules, DataWeave can also load and translate declarations from a custom module file into DataWeave type directives that can be accessed in the same way as types from any other DataWeave module.
//PR for this doc: https://github.com/mulesoft/docs-dataweave/pull/224/files

== New DataWeave Function Modules and Features

DataWeave introduces the following changes:

> * xref:dw-dtd.adoc[] module is new and includes the following:

Author: ```suggestion
* xref:dw-dtd.adoc[] module (`dw::xml::Dtd`) is new and includes the following:
```

//          advanced scenario where typed parameters, or generics, will now be able
//          to be declared at the function call level.  So one can say
//          myFunc<String>()  to signal that the generic in myFunc<T>() should
//          be a String. In the past, generics were always inferred which led
//          to some type checking errors. Because this required a syntax change,
//          it will only be available to scripts declaring %dw 2.5  or higher
//          (in the future) as their desired version.
//github MERGE (type parameter application):
//             https://github.com/mulesoft/docs-dataweave/pull/215/files

** Type selector expression //TODO: Clarify.
//From ANA: Type selector expression (@Martín Cousido): Users will be able to
//          navigate through type definitions to define DW types, so from a
//          complex type you can select a nested part as a new definition.
// * See Google doc listed in Slack at https://salesforce-internal.slack.com/archives/C011SNL2469/p1672164627375259: https://docs.google.com/document/d/1KLNh_FDnRIPG_nC4IRzBnN2Pr79d0tuvI6shZ12EcKc/edit#heading=h.epfq33bmp8h8

== New DataWeave Function Modules and Features

> DataWeave introduces the following:

Author: ```suggestion
DataWeave introduces the following changes:
```

* xref:dw-dtd.adoc[] module
** xref:dw-dtd-functions-doctypeasstring.adoc[] function 
** xref:dw-dtd-types.adoc[] types

> 

Author: TODO: How were all these modified?

Author: done

The following functions are updated or new in this release:

* xref:dw-binaries.adoc[] updates: 

> ** xref:dw-binaries-functions-frombase64.adoc[]

Author: ```suggestion
```
Non-substantive change

> ** xref:dw-binaries-functions-fromhex.adoc[]

Author: ```suggestion
```
Non-substantive change

> ** xref:dw-binaries-functions-tobase64.adoc[]

Author: ```suggestion
```
Non-substantive change

* xref:dw-binaries-functions-concatwith.adoc[] (new in 2.5)

* xref:dw-coercions.adoc[] updates:

> ** xref:dw-coercions-functions-toarray.adoc[]

Author: ```suggestion
```

> ** xref:dw-coercions-functions-tobinary.adoc[]

Author: ```suggestion
```

> ** xref:dw-coercions-functions-todate.adoc[]

Author: ```suggestion
```

> ** xref:dw-coercions-functions-todatetime.adoc[]

Author: ```suggestion
```

> ** xref:dw-coercions-functions-tolocaldatetime.adoc[]

Author: ```suggestion
```

> ** xref:dw-coercions-functions-tolocaltime.adoc[]

Author: ```suggestion
```

> ** xref:dw-coercions-functions-tonumber.adoc[]

Author: ```suggestion
```

> ** xref:dw-coercions-functions-tostring.adoc[]

Author: ```suggestion
** xref:dw-coercions-functions-tostring.adoc[] adds the `locale` parameter.
```

Author: This is the only substantive change

> ** xref:dw-coercions-functions-totime.adoc[]

Author: ```suggestion
```

* xref:dw-core.adoc[] updates:

> ** xref:dw-core-annotations.adoc[]

Author: ```suggestion
** xref:dw-core-annotations.adoc[]: `@UntrustedCode&#40;&#41;` changes to `@UntrustedCode&#40;privileges: Array<String&#62;&#41;`
```

> ** xref:dw-core-functions-distinctby.adoc[]

Author: ```suggestion
```

> ** xref:dw-core-functions-filter.adoc[]

Author: ```suggestion
```

> ** xref:dw-core-functions-filterobject.adoc[]

Author: ```suggestion
```

> ** xref:dw-core-functions-groupby.adoc[]

Author: TODO CHK: 

Signature changes from 
`groupBy<T, R&#62;&#40;items: Array<T&#62;, criteria: &#40;item: T, index: Number&#41; &#45;&#62; R&#41;: { &#40;R&#41;: Array<T&#62; } `

to this

`groupBy<T, R&#62;&#40;items: Array<T&#62;, criteria: &#40;item: T, index: Number&#41; &#45;&#62; R&#41;: {&#124; &#40;R&#41;: Array<T&#62; &#124;}`

> ** xref:dw-core-functions-minusminus.adoc[]

Author: ```suggestion
```

> ** xref:dw-core-functions-typeof.adoc[]

Author: ```suggestion
```

> * xref:dw-crypto.adoc[] update:

Author: ```suggestion
```

> ** xref:dw-crypto-functions-hmacwith.adoc[]

Author: ```suggestion
```

* xref:dw-dates.adoc[] updates:

> ** xref:dw-dataformat-types.adoc[]

Author: TODO: check xref:dw-dataformat-types.adoc[]

> ** xref:dw-dates-functions-atbeginningofday.adoc[]

Author: ```suggestion
```

> ** xref:dw-dates-functions-atbeginningofhour.adoc[]

Author: ```suggestion
```

> ** xref:dw-dates-functions-atbeginningofmonth.adoc[]

Author: ```suggestion
```

> ** xref:dw-dates-functions-atbeginningofweek.adoc[]

Author: ```suggestion
```

> ** xref:dw-dates-functions-atbeginningofyear.adoc[]

Author: ```suggestion
```

> ** xref:dw-dates-functions-date.adoc[]

Author: ```suggestion
```

> ** xref:dw-dates-functions-datetime.adoc[]

Author: ```suggestion
```

> ** xref:dw-dates-functions-localdatetime.adoc[]

Author: ```suggestion
```

> ** xref:dw-dates-functions-localtime.adoc[]

Author: ```suggestion
```

> ** xref:dw-dates-functions-time.adoc[]

Author: ```suggestion
```

> ** xref:dw-dates-functions-today.adoc[]

Author: ```suggestion
```

> * xref:dw-math.adoc[] updates:

Author: ```suggestion
```

> ** xref:dw-math-functions-acos.adoc[]

Author: ```suggestion
```

> ** xref:dw-math-functions-asin.adoc[]

Author: ```suggestion
```

> ** xref:dw-math-functions-atan.adoc[]

Author: ```suggestion
```

> ** xref:dw-math-functions-cos.adoc[]

Author: ```suggestion
```

> ** xref:dw-math-functions-log10.adoc[]

Author: ```suggestion
```

> ** xref:dw-math-functions-logn.adoc[]

Author: ```suggestion
```

> ** xref:dw-math-functions-sin.adoc[]

Author: ```suggestion
```

> ** xref:dw-math-functions-tan.adoc[]

Author: ```suggestion
```

> ** xref:dw-math-functions-todegrees.adoc[]

Author: ```suggestion
```

> ** xref:dw-math-functions-toradians.adoc[]

Author: ```suggestion
```

> ** xref:dw-math-variables.adoc[]

Author: ```suggestion
```

> * xref:dw-multipart.adoc[] updates:

Author: ```suggestion
```

> ** xref:dw-multipart-functions-field.adoc[]

Author: ```suggestion
```

> ** xref:dw-multipart-functions-file.adoc[]

Author: ```suggestion
```

> ** xref:dw-multipart-functions-form.adoc[]

Author: ```suggestion
```

> * xref:dw-runtime.adoc[] updates:

Author: ```suggestion
* xref:dw-runtime.adoc[] update:
```

> ** xref:dw-runtime-functions-eval.adoc[]

Author: ```suggestion
```

> ** xref:dw-runtime-types.adoc[]

Author: TODO: double check for any changes to
 xref:dw-runtime-types.adoc[]

** xref:dw-runtime-functions-version.adoc[] (new in 2.5)

> * xref:dw-periods.adoc[] updates:

Author: ```suggestion
```

> ** xref:dw-periods-functions-days.adoc[]

Author: ```suggestion
```

> ** xref:dw-periods-functions-duration.adoc[]

Author: ```suggestion
```

> ** xref:dw-periods-functions-hours.adoc[]

Author: ```suggestion
```

> ** xref:dw-periods-functions-minutes.adoc[]

Author: ```suggestion
```

> ** xref:dw-periods-functions-months.adoc[]

Author: ```suggestion
```

> ** xref:dw-periods-functions-period.adoc[]

Author: ```suggestion
```

> ** xref:dw-periods-functions-seconds.adoc[]

Author: ```suggestion
```

> ** xref:dw-periods-functions-years.adoc[]

Author: ```suggestion
```

> * xref:dw-strings.adoc[] updates:

Author: ```suggestion
```

> ** xref:dw-strings-functions-countcharactersby.adoc[]

Author: ```suggestion
```

> ** xref:dw-strings-functions-countmatches.adoc[]

Author: ```suggestion
```

> ** xref:dw-strings-functions-everycharacter.adoc[]

Author: ```suggestion
```

> ** xref:dw-strings-functions-first.adoc[]

Author: ```suggestion
```

> ** xref:dw-strings-functions-last.adoc[]

Author: ```suggestion
```

> ** xref:dw-strings-functions-somecharacter.adoc[]

Author: ```suggestion
```

> ** xref:dw-strings-functions-substringby.adoc[]

Author: ```suggestion
```

> ** xref:dw-strings-functions-withmaxsize.adoc[]

Author: ```suggestion
** xref:dw-strings-functions-withmaxsize.adoc[]
```

= What's New in DataWeave

DataWeave 2.5.0 is bundled with the Mule 4.5.0 release. For details about the Mule release, see
xref:mule-runtime/mule-4.5.0-release-notes.adoc[Mule Runtime Engine 4.5.0 Release Notes].

//INCLUDE SHARED CONTENT with release notes

> include::release-notes::partial$dataweave/dataweave-whats-new-2.5.0.adoc[tag=dataweaveFeatures250]

Author: works locally for me but I haven't merged that include yet

Author: ty
